[JSC] Add `JSAsyncFromSyncIterator`

sosukesuzuki opened the pull request:
#### 1051b2abe26108a39437cddeabf4a004cbb74ad6
<pre>
[JSC] Add `JSAsyncFromSyncIterator`
<a href="https://bugs.webkit.org/show_bug.cgi?id=279251">https://bugs.webkit.org/show_bug.cgi?id=279251</a>

Reviewed by NOBODY (OOPS!).

This patch adds `JSAsyncFromSyncIterator` using `JSInternalFieldObjectImpl`.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/AsyncFromSyncIteratorPrototype.js:
(return):
(throw):
(linkTimeConstant.createAsyncFromSyncIterator):
(linkTimeConstant.AsyncFromSyncIterator): Deleted.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapSubspaceTypes.h:
* Source/JavaScriptCore/runtime/JSAsyncFromSyncIterator.cpp: Added.
(JSC::JSAsyncFromSyncIterator::createWithInitialValues):
(JSC::JSAsyncFromSyncIterator::finishCreation):
(JSC::JSAsyncFromSyncIterator::visitChildrenImpl):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSAsyncFromSyncIterator.h: Added.
* Source/JavaScriptCore/runtime/JSAsyncFromSyncIteratorInlines.h: Added.
(JSC::JSAsyncFromSyncIterator::createStructure):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::asyncFromSyncIteratorPrototype const):
(JSC::JSGlobalObject::asyncFromSyncIteratorStructure const):
</pre>

Constellation reviewed (changes requested):
bind is costly operation, so it is not fine. You need to piggy back this as a context somehow.

sosukesuzuki committed:
[JSC] Close the sync iterator when rejected promise is yielded in AsyncFromSyncIterator
https://bugs.webkit.org/show_bug.cgi?id=273768

Reviewed by NOBODY (OOPS!).

This patch implements the new normative change[1].
When the AsyncFromSyncIterator yields a rejected promise, it calls the return method of the
underlying syncIterator and closes it.

[1]: https://github.com/tc39/ecma262/pull/2600

* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/builtins/AsyncFromSyncIteratorPrototype.js:
(linkTimeConstant.asyncFromSyncIteratorOnRejected):
(return):
(throw):
(linkTimeConstant.AsyncFromSyncIterator):

sosukesuzuki commented on Source/JavaScriptCore/builtins/AsyncFromSyncIteratorPrototype.js:
```diff
@@ -180,4 +199,9 @@ function AsyncFromSyncIterator(syncIterator, nextMethod)
 
     @putByIdDirectPrivate(this, "syncIterator", syncIterator);
     @putByIdDirectPrivate(this, "nextMethod", nextMethod);
+    @putByIdDirectPrivate(this, "onRejected", function (error, promise) {
+        "use strict";
+
+        @asyncFromSyncIteratorOnRejected(syncIterator, error, promise);
+    });
```
@Constellation  I've fixed to create a closure only once inside AsyncFromSyncIterator constructor. What do you think?
